### ISSUE TITLE
Support any OAuth server with custom config

### ISSUE THREAD

**Issue by hickford** [git-ecosystem/git-credential-manager#1047]
To add OAuth support for a new forge requires around a thousand lines of code which has to be reviewed (example PRs: [GitLab](https://github.com/GitCredentialManager/git-credential-manager/pull/591), [Gitea](https://github.com/GitCredentialManager/git-credential-manager/pull/879), [Gitee](https://github.com/GitCredentialManager/git-credential-manager/pull/1000)). Each new assembly adds megabytes to git-credential-manager binary size. Is any code necessary? All are similar OAuth implementations with only trivial differences in OAuth parameters.

Ideally you could use GCM with any generic OAuth forge with only simple config to specify the OAuth parameters, no code change necessary.

https://github.com/hickford/git-credential-oauth has this feature. Here's an example config for gitee.com:

```ini
[credential "https://gitee.com"]
	oauthClientId = b5da229f85d94f5d728c43e429c69527fe2dcc387d4ffcb677b91c93983adaa3
	oauthClientSecret = a9f88c722f66e54184204efa0f49b960f2666947e0b14bd0d73a8c9c1c33c340
	oauthScopes = projects gists
	oauthAuthURL = /oauth/authorize
	oauthTokenURL = /oauth/token
	# workaround server confused by random ports
	oauthRedirectURL = http://127.0.0.1:53119
```

**Comment by maikebing** [comment]
Great job！

**Comment by ldennington** [comment]
Thanks @hickford! This is something we've been talking about internally for awhile given the number of different Git hosts/forges that want GCM support via OAuth. In light of this, we've decided generic OAuth provider support would make sense here. We plan to have this complete in time for Git 2.40 release.

**Comment by mjcheetham** [comment]
Now that #1062 has been merged this means the next GCM release will include support for generic OAuth providers!

**Comment by Kongpcmail** [comment]
> To add OAuth support for a new forge requires around a thousand lines of code which has to be reviewed (example PRs: [GitLab](https://github.com/GitCredentialManager/git-credential-manager/pull/591), [Gitea](https://github.com/GitCredentialManager/git-credential-manager/pull/879), [Gitee](https://github.com/GitCredentialManager/git-credential-manager/pull/1000)). Each new assembly adds megabytes to git-credential-manager binary size. Is any code necessary? All are similar OAuth implementations with only trivial differences in OAuth parameters.
> 
> Ideally you could use GCM with any generic OAuth forge with only simple config to specify the OAuth parameters, no code change necessary.
> 
> https://github.com/hickford/git-credential-oauth has this feature. Here's an example config for gitee.com:
> 
> ```ini
> [credential "https://gitee.com"]
> 	oauthClientId = b5da229f85d94f5d728c43e429c69527fe2dcc387d4ffcb677b91c93983adaa3
> 	oauthClientSecret = a9f88c722f66e54184204efa0f49b960f2666947e0b14bd0d73a8c9c1c33c340
> 	oauthScopes = projects gists
> 	oauthAuthURL = /oauth/authorize
> 	oauthTokenURL = /oauth/token
> 	# workaround server confused by random ports
> 	oauthRedirectURL = http://127.0.0.1:53119
> ```

**Comment by maikebing** [comment]
Maybe the LOGO path configuration is missing?  @mjcheetham

**Comment by mjcheetham** [comment]
> Maybe the LOGO path configuration is missing? @mjcheetham

Are you looking for a way to add custom branding/logos to the generic OAuth UI prompts?

<img width="532" alt="image" src="https://user-images.githubusercontent.com/5658207/217380210-afc48385-90bb-4351-be7c-a1da00775301.png">

At the moment we don't plan to add support for customising the UI with branding. If you wanted to submit a PR adding the functionality we would consider it! Configuration pointing at either a local or remote file (URL) may be one acceptable method for this, but we'd want to be sure pulling external content from the Internet to be rendered was safe.

**Comment by ldennington** [comment]
Closing as this feature was released in 2.0.935 🎉 .
